### PR TITLE
Multi-namespace generation

### LINKED ISSUE #349: failing validation of matnwb-generated files with cached namespace
matnwb-generated NWB files fail validation due to missing experimental.yaml file. The following replicates the error:

```matlab
nwb = NwbFile( ...
    'session_description', 'mouse in open exploration',...
    'identifier', 'Mouse5_Day3', ...
    'session_start_time', datetime(2018, 4, 25, 2, 30, 3) ...
);
nwbExport(nwb, 'simple_file_export.nwb')
```

```bash
$  python -m pynwb.validate simple_file_export.nwb
```
full traceback:
```python
Traceback (most recent call last):
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/pynwb/validate.py", line 136, in <module>
    main()
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/pynwb/validate.py", line 72, in main
    ns_deps = NWBHDF5IO.load_namespaces(catalog, path)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/utils.py", line 583, in func_call
    return func(args[0], **pargs)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/backends/hdf5/h5tools.py", line 147, in load_namespaces
    return cls.__load_namespaces(namespace_catalog, namespaces, open_file_obj)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/backends/hdf5/h5tools.py", line 181, in __load_namespaces
    d.update(namespace_catalog.load_namespaces(cls.__ns_spec_path, reader=reader))
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/utils.py", line 583, in func_call
    return func(args[0], **pargs)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/spec/namespace.py", line 538, in load_namespaces
    ret[ns['name']] = self.__load_namespace(ns, reader, resolve=resolve)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/spec/namespace.py", line 448, in __load_namespace
    self.__load_spec_file(reader, s['source'], catalog, types_to_load=types_to_load, resolve=resolve)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/spec/namespace.py", line 393, in __load_spec_file
    d = reader.read_spec(spec_source)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/backends/hdf5/h5_utils.py", line 331, in read_spec
    return self.__read(spec_path)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/backends/hdf5/h5_utils.py", line 320, in __read
    s = self.__group[path][()]
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/h5py/_hl/group.py", line 305, in __getitem__
    oid = h5o.open(self.id, self._e(name), lapl=self._lapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5o.pyx", line 190, in h5py.h5o.open

```

No issue when the pynwb validator is is run with the  `--no-cached-namespace`
There's also no issue with pynwb-generated files.

### PR DESCRIPTION
**Pull request by ln-vidrio**
Fixes #349

### CODE REVIEW

**Comment by cechava** [comment]
thanks for the pointer. it makes sense. looks like the changes are creating errors when reading the files back in

**Comment by cechava** [comment]
It also looks like the original issue with validation persists. The same error is output by the pynwb.validate function:
```python
KeyError: "Unable to open object (object 'experimental.yaml' doesn't exist)"
```

**Comment by cechava** [comment]
I noticed that in the `+schemes/exportJson.m` function we weren't stripping the .yaml extension from the sources. I updated the function hoping that it would resolve the issue with the file validation. However, the issue persists. Committed the change anyways.

**Review by cechava — COMMENTED** [review]
Thanks for fixing the reading errors! The generated files are still failing validation though. I'm trying to figure out why that is. I suspect it has something to do with the fact that the experimental.yaml file is under the 'common' folder for the Python hdmf package. I'm trying to nail down the issue. 
To replicate: 
```matlab
nwb = NwbFile( ...
    'session_description', 'mouse in open exploration',...
    'identifier', 'Mouse5_Day3', ...
    'session_start_time', datetime(2018, 4, 25, 2, 30, 3) ...
);
nwbExport(nwb, 'simple_file_export.nwb')
```
```bash
$ python -m pynwb.validate simple_file_export.nwb 
Traceback (most recent call last):
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/pynwb/validate.py", line 136, in <module>
    main()
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/pynwb/validate.py", line 72, in main
    ns_deps = NWBHDF5IO.load_namespaces(catalog, path)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/utils.py", line 583, in func_call
    return func(args[0], **pargs)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/backends/hdf5/h5tools.py", line 150, in load_namespaces
    return cls.__load_namespaces(namespace_catalog, namespaces, open_file_obj)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/backends/hdf5/h5tools.py", line 187, in __load_namespaces
    d.update(namespace_catalog.load_namespaces(cls.__ns_spec_path, reader=reader))
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/utils.py", line 583, in func_call
    return func(args[0], **pargs)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/spec/namespace.py", line 544, in load_namespaces
    ret[ns['name']] = self.__load_namespace(ns, reader, resolve=resolve)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/spec/namespace.py", line 451, in __load_namespace
    self.__load_spec_file(reader, s['source'], catalog, types_to_load=types_to_load, resolve=resolve)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/spec/namespace.py", line 393, in __load_spec_file
    d = reader.read_spec(spec_source)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/backends/hdf5/h5_utils.py", line 333, in read_spec
    return self.__read(spec_path)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/backends/hdf5/h5_utils.py", line 322, in __read
    s = self.__group[path][()]
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/h5py/_hl/group.py", line 305, in __getitem__
    oid = h5o.open(self.id, self._e(name), lapl=self._lapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5o.pyx", line 190, in h5py.h5o.open
KeyError: "Unable to open object (object 'experimental' doesn't exist)"
```

**Comment by ln-vidrio** [comment]
It looks like it's looking for an object called "experimental" even though the namespace is specified as `hdmf-experimental` in its namespace.yaml file. Not sure what that's supposed to mean unless there's even more hidden syntax in the spec.

**Comment by cechava** [comment]
Doing some printing, I found that the `pynwb.validate` function is looking for the 'experimental' object in the following HDF5 group:
`<KeysViewHDF5 ['base', 'namespace', 'sparse', 'table']>`
The issue seems to be that the cached spec for experimental contains both the 'hdmf-common' and the 'hdmf-experimental' namespaces and the validator is only looking at the first container. 
The solution would be to modify the `+scheme/exportJson.m` function to only cache the relevant namespace. 
Let me take a stab at it.

**Comment by cechava** [comment]
latest commit solves that error with solver. Now getting another error:
```python
Traceback (most recent call last):
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/pynwb/validate.py", line 136, in <module>
    main()
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/pynwb/validate.py", line 72, in main
    ns_deps = NWBHDF5IO.load_namespaces(catalog, path)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/utils.py", line 583, in func_call
    return func(args[0], **pargs)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/backends/hdf5/h5tools.py", line 150, in load_namespaces
    return cls.__load_namespaces(namespace_catalog, namespaces, open_file_obj)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/backends/hdf5/h5tools.py", line 187, in __load_namespaces
    d.update(namespace_catalog.load_namespaces(cls.__ns_spec_path, reader=reader))
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/utils.py", line 583, in func_call
    return func(args[0], **pargs)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/spec/namespace.py", line 548, in load_namespaces
    ret[ns['name']] = self.__load_namespace(ns, reader, resolve=resolve)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/spec/namespace.py", line 452, in __load_namespace
    self.__load_spec_file(reader, s['source'], catalog, types_to_load=types_to_load, resolve=resolve)
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/spec/namespace.py", line 402, in __load_spec_file
    temp_dict = {k: None for k in __reg_spec(self.__group_spec_cls, spec_dict)}
  File "/Users/cesar/anaconda3/envs/pynwb-dev/lib/python3.9/site-packages/hdmf/spec/namespace.py", line 380, in __reg_spec
    dt_def = spec_dict.get(spec_cls.def_key())
AttributeError: 'list' object has no attribute 'get'
```

We are getting close to this working properly

**Comment by rly** [comment]
For context, here is the result of PyNWB's namespace caching:
image

The `/specifications/hdmf-experimental/0.1.0/namespace` dataset contains a JSON object:
```json
{
   "namespaces":[
      {
         "doc":"Experimental data structures provided by HDMF. These are not guaranteed to be available in the future",
         "schema":[
            {
               "namespace":"hdmf-common"
            },
            {
               "source":"experimental"
            },
            {
               "source":"resources"
            }
         ],
         "name":"hdmf-experimental",
         "full_name":"HDMF Experimental",
         "version":"0.1.0",
		 .....
      }
   ]
}
```
The `source` values are the names of the datasets under the `[namespace_name]/[namespace_version]` group.

**Comment by cechava** [comment]
@rly is there a straightforward way to output one of those JSON object s from an NWB file? I would like to compare what is included and how in a pynwb-generated vs matnwb-generated file. I'm using HDFView on mac, but it just lets me see a long string that is hard to parse by eye.

**Comment by ln-vidrio** [comment]
I just use an online JSON prettifier (at my own peril). I'm sure there's some offline tool that can do that as well. I unfortunately don't see that much of a difference between the snippet our own embedded spec.

**Comment by rly** [comment]
I also use an online JSON prettifier, but for reference, in Python, you can access the JSON object using:
```python
import h5py
import json
from pprint import pprint
f = h5py.File(r"C:\Users\Ryan\Documents\NWB_Data\GiocomoData_dandiset53\000053\sub-npI1\sub-npI1_ses-20190413_behavior+ecephys.nwb", "r")
pprint(json.loads(f["/specifications/hdmf-experimental/0.1.0/namespace"][()]))
```

**Comment by cechava** [comment]
The issue is with `/specifications/hdmf-experimental/0.1.0/experimental` and matnwb. The `experimental.yaml` schema file defines an empty HDF5 group. Pynwb doesn't cache this empty HDF5 group to the file.

```python
import h5py
import json
from pprint import pprint
f = h5py.File('test_file_pynwb.nwb', "r")
pprint(json.loads(f["/specifications/hdmf-experimental/0.1.0/experimental"][()]))

{'datasets': [{'attributes': [{'doc': 'Reference to the VectorData object that '
                                      'contains the enumerable elements',
                               'dtype': {'reftype': 'object',
                                         'target_type': 'VectorData'},
                               'name': 'elements'}],
               'data_type_def': 'EnumData',
               'data_type_inc': 'VectorData',
               'doc': 'Data that come from a fixed set of values. A data value '
                      'of i corresponds to the i-th value in the VectorData '
                      "referenced by the 'elements' attribute.",
               'dtype': 'uint8'}]}
```
 In contrast, matnwb does and the `pynwb.validate` function doesn't know how to handle it
 
```python
f = h5py.File('simple_file_export.nwb', "r")
pprint(json.loads(f["/specifications/hdmf-experimental/0.1.0/experimental"][()]))
{'datasets': [{'attributes': [{'doc': 'Reference to the VectorData object that '
                                      'contains the enumerable elements',
                               'dtype': {'reftype': 'object',
                                         'target_type': 'VectorData'},
                               'name': 'elements'}],
               'data_type_def': 'EnumData',
               'data_type_inc': 'VectorData',
               'doc': 'Data that come from a fixed set of values. A data value '
                      'of i corresponds to the i-th value in the VectorData '
                      "referenced by the 'elements' attribute.",
               'dtype': 'uint8'}],
 'groups': [[]]}
```

The most straightforward thing to do is to put a check somewhere so that empty groups in the namespace are not cached to file.

**Comment by bendichter** [comment]
@cechava you are correct, nice find! A little terminology nitpick: The [schema language](https://schema-language.readthedocs.io/en/stable/specification_language_description.html) defines data primitives (Group, Dataset, Attributes and Links) agnostic to the data storage backend. These could technically be written to another backend like Zarr or a relational database. In practice we usually use HDF5, so these groups will probably end up being HDF5 Groups when stored to disk, but the issue is in reading this schema, so they are not HDF5 groups _yet_, only Groups. It looks like the issue is that the root Groups list of the namespace is an empty list.

@rly do you think we could remove the empty "groups" from `experimental.yaml`?

**Comment by cechava** [comment]
Latest commit produces files that pass pynwb validation. @ln-vidrio Let me know if it looks good.

**Review by cechava — APPROVED** [review]
All looks good to me. Let me know if my changes look good to you or if you foresee problems with my approach.
